app/src/test/java/de/westnordost/streetcomplete/data/elementfilter/ElementFiltersParserTest.kt

```diff
@@ -497,6 +501,134 @@ class ElementFiltersParserTest {
+        notMatchesTags(mapOfKeys("a", "b", "c", "d"), expr)
+    }
+
+    @Test fun `nested not with or and switched operands`() {
+        val expr = "!(c or !(a and b))" // equals a and b and !(c)
+        matchesTags(mapOfKeys("a", "b"), expr)
+        matchesTags(mapOfKeys("a", "b", "d"), expr)
+        notMatchesTags(mapOfKeys("a"), expr)
+        notMatchesTags(mapOfKeys("c"), expr)
+        notMatchesTags(mapOfKeys("b", "c"), expr)
+        notMatchesTags(mapOfKeys("a", "b", "c"), expr)
+        notMatchesTags(mapOfKeys("a", "b", "c", "d"), expr)
+    }
+
+    @Test
+    fun `test expression parsing AddCycleway`() {
+        val residentialWayIn30Zone = way(tags = mapOf(
+            HIGHWAY to RESIDENTIAL,
+            "maxspeed" to "30",
+            "zone:traffic" to "DE:urban",
+            "zone:maxspeed" to "DE:30",
+        ))
+
+        val residentialWayInBuiltUpAreaWithMaxspeed30 = way(tags = mapOf(
+            HIGHWAY to RESIDENTIAL,
+            "maxspeed" to "30",
+            "zone:traffic" to "DE:urban"
+        ))
+
+        val notInZone30OrLess = """
+   !(~"${(MAXSPEED_TYPE_KEYS + "maxspeed").joinToString("|")}" ~ ".*:(30|20)")
+"""
+
+        val notInZone30OrLessOld = """
+   ~"${(MAXSPEED_TYPE_KEYS + "maxspeed").joinToString("|")}" ~ ".*:(urban|rural|trunk|motorway|nsl_single|nsl_dual)"
+"""
+
+        val expression = getCycleWayQuestFilterExpression(notInZone30OrLess)
+        val oldExpression = getCycleWayQuestFilterExpression(notInZone30OrLessOld)
+
+        // using the ! operator makes it possible to filter out more elements
+        assertFalse("residentialWayIn30Zone was matched", expression.matches(residentialWayIn30Zone))
+        assertTrue("residentialWayIn30Zone wasn't matched", oldExpression.matches(residentialWayIn30Zone))
+
+        assertTrue("residentialWayInBuiltUpAreaWithMaxspeed30 wasn't matched", expression.matches(residentialWayInBuiltUpAreaWithMaxspeed30))
+        assertTrue("residentialWayInBuiltUpAreaWithMaxspeed30 wasn't matched", oldExpression.matches(residentialWayInBuiltUpAreaWithMaxspeed30))
+    }
+
+    private fun getCycleWayQuestFilterExpression(notInZone30OrLessNew: String): ElementFilterExpression {
+        val expression = """
+        ways with (
+            highway ~ primary|primary_link|secondary|secondary_link|tertiary|tertiary_link|unclassified
+            or highway = residential and (maxspeed > 33 or $notInZone30OrLessNew)
+          )
+          and !cycleway
+          and !cycleway:left
+          and !cycleway:right
+          and !cycleway:both
+          and !sidewalk:bicycle
+          and !sidewalk:left:bicycle
+          and !sidewalk:right:bicycle
+          and !sidewalk:both:bicycle
+          and (
+            !maxspeed
+            or maxspeed > 20
+            or $notInZone30OrLessNew
+          )
+          and surface !~ ${ANYTHING_UNPAVED.joinToString("|")}
+    """.toElementFilterExpression()
+        return expression
+    }
+
```

Comment:
This looks out of place for me. Why would this test be necessary when you already test every variant? If you really think that this test makes sense, it should go into the cycleway quest.

Otherwise, concise and easy to read tests!